### PR TITLE
Add topic.json_upload/import

### PR DESCRIPTION
**Pull request by reiterl**
Add some tests.

### CODE REVIEW

**Comment by reiterl** [comment]
The cov of the feature/remove-template-fields is too low.

File: openslides_backend/action/actions/topic/mixins.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
+from ....shared.filters import FilterOperator
+from ...action import Action
+
+
+class DuplicateCheckMixin(Action):
+    def init_duplicate_set(self, meeting_id: int) -> None:
+        self.all_titles_in_meeting = set(
+            [
```
Comment (jsangmeister) [inline openslides_backend/action/actions/topic/mixins.py:8]: remove the array brackets

File: openslides_backend/action/actions/topic/json_upload.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+    def update_instance(self, instance: Dict[str, Any]) -> Dict[str, Any]:
+        data = instance.pop("data")
+
+        # enrich data with meeting_id
+        for entry in data:
+            entry["meeting_id"] = instance["meeting_id"]
+
+        # generate headers
+        self.headers = [
```
Comment (jsangmeister) [inline openslides_backend/action/actions/topic/json_upload.py:63]: Since these are constant, define them as a class variable

File: tests/system/action/topic/test_json_upload.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        assert result["headers"][0] == {
+            "property": "title",
+            "type": "string",
+        }
+        assert result["rows"][0] == {
+            "status": "new",
+            "error": [],
+            "data": {"title": "test", "meeting_id": 22},
+        }
+        assert result["statistics"] == {"itemCount": 1, "Created": 1, "Error": 0}
```
Comment (jsangmeister) [inline tests/system/action/topic/test_json_upload.py:97]: Just assert the full result at once to make sure no unwanted data is included

File: tests/system/action/topic/test_json_upload.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        response = self.request(
+            "topic.json_upload",
+            {
+                "meeting_id": 22,
+                "data": [{"title": "test"}, {"title": "bla"}, {"title": "test"}],
+            },
+        )
+        self.assert_status_code(response, 200)
+        result = response.json["results"][0][0]
+        assert result["rows"][2]["error"] == ["Duplicate"]
```
Comment (jsangmeister) [inline tests/system/action/topic/test_json_upload.py:124]: assert that the status is `error`. Also assert that the other 2 entries are imported correctly

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/topic/json_upload.py (hunk 2 above)
Comment (reiterl) [reply openslides_backend/action/actions/topic/json_upload.py:63]: Done.

File: openslides_backend/action/actions/topic/import_.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+                        fields={
+                            "id": store_id,
+                            "result": None,
+                        },
+                    )
+                ],
+                user_id=self.user_id,
+                locked_fields={},
+            )
+        )
```
Comment (jsangmeister) [inline openslides_backend/action/actions/topic/import_.py:65]: Remove this: if the import fails for some reason, we need to have this data still available.

File: tests/system/action/topic/test_json_upload.py (hunk 3 above)
Comment (reiterl) [reply tests/system/action/topic/test_json_upload.py]: Done.

File: tests/system/action/topic/test_json_upload.py (hunk 4 above)
Comment (reiterl) [reply tests/system/action/topic/test_json_upload.py:131]: Done. Check the status and check the rows in the action worker.

File: openslides_backend/action/actions/topic/json_upload.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        for entry in self.rows:
+            itemCount += 1
+            if entry["status"] == "new":
+                itemNew += 1
+            if entry["status"] == "error":
+                itemError += 1
+        self.statistics = {
+            "itemCount": itemCount,
+            "Created": itemNew,
+            "Error": itemError,
```
Comment (jsangmeister) [inline openslides_backend/action/actions/topic/json_upload.py:84]: Do I understand it correctly that these strings are to be displayed as-is in the client? If so, we should choose more suitable ones.

File: openslides_backend/action/actions/topic/json_upload.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        fqid = fqid_from_collection_and_id("action_worker", self.new_store_id)
+        self.datastore.write_action_worker(
+            WriteRequest(
+                events=[
+                    Event(
+                        type=EventType.Create,
+                        fqid=fqid,
+                        fields={
+                            "id": self.new_store_id,
+                            "result": self.rows,
```
[the remark this answers is not in the document]
Comment (r-peschke) [reply openslides_backend/action/actions/topic/json_upload.py:98]: It is not necessary, because there is a type-specific action, i.e. topic.import, for this

[the remark this answers is not in the document]
Comment (jsangmeister) [reply openslides_backend/action/actions/topic/json_upload.py:98]: Yes, and `topic.import` only gets an action worker id. If in the action worker it is saved to which import it belongs, the validation is easier that no invalid id was provided.

File: openslides_backend/action/actions/topic/import_.py (hunk 5 above)
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py:65]: Mmh, this also handles the abort. Should an abort remove the data or not?

File: openslides_backend/action/actions/topic/import_.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+class TopicImport(DuplicateCheckMixin, Action):
+    """
+    Action to import a result from the action_worker.
+    """
+
+    model = ActionWorker()
+    schema = DefaultSchema(ActionWorker()).get_default_schema(
+        additional_required_fields={
+            "id": required_id_schema,
+            "command": {"type": "string", "enum": ["import", "abort"]},
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py:27]: Okay, removed the command and the abort.

File: openslides_backend/action/actions/topic/import_.py (hunk 5 above)
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py:65]: As I see in the later comments, we want to remove abort.

File: openslides_backend/action/actions/topic/import_.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        if command == "import":
+            meeting_id = self.get_meeting_id(instance)
+            self.init_duplicate_set(meeting_id)
+            worker = self.datastore.get(
+                fqid_from_collection_and_id("action_worker", store_id), ["result"]
+            )
+            action_payload = [
+                entry["data"]
+                for entry in worker.get("result", [])
+                if entry["status"] == "new"
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py:44]: That will lead to such errors:
TypeError: Object of type ImportStatus is not JSON serializable

File: openslides_backend/action/actions/topic/import_.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        )
+        if worker.get("result") and isinstance(worker["result"], list):
+            meeting_id = None
+            okay = True
+            for entry in worker["result"]:
+                if "status" not in entry or "error" not in entry or "data" not in entry:
+                    okay = False
+                meeting_id = entry["data"].get("meeting_id")
+            if okay and meeting_id:
+                return meeting_id
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py:87]: Updated.

File: openslides_backend/action/actions/topic/import_.py [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        raise ActionException("Topic import is aborted or done.")
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py:88]: Use a "import": "topic" flag, to see, it's data are right.

File: openslides_backend/action/actions/topic/json_upload.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/json_upload.py:98]: I think it is a good idea, to include an import: "topic" in the action_worker, so that the import action can look, that has correct data. I include it.

File: openslides_backend/action/actions/topic/import_.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py:44]: Add an enum `ImportStatus` and use it in the actions and the tests.

File: openslides_backend/action/actions/topic/json_upload.py (hunk 6 above)
Comment (reiterl) [reply openslides_backend/action/actions/topic/json_upload.py:84]: Use Total/New/Error now. Or do you have a better idea?

File: openslides_backend/action/actions/topic/import_.py (hunk 5 above)
Comment (r-peschke) [reply openslides_backend/action/actions/topic/import_.py]: > Remove this: if the import fails for some reason, we need to have this data still available.

IMO we should remove the data, if the import is successful. The topic import never will be large, but others will.

File: openslides_backend/action/actions/topic/json_upload.py (hunk 6 above)
Comment (r-peschke) [reply openslides_backend/action/actions/topic/json_upload.py]: Use Luisa's, see https://github.com/OpenSlides/openslides-backend/issues/1661#issuecomment-1466109237, like
* total: 10
* created: 5
* omitted: 5

File: tests/system/action/topic/test_json_upload.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+    def test_json_upload_results(self) -> None:
+        response = self.request(
+            "topic.json_upload",
+            {"meeting_id": 22, "data": [{"title": "test"}]},
+        )
+        self.assert_status_code(response, 200)
+        self.assert_model_exists(
+            "action_worker/1",
+            {
+                "result": {
```
Comment (r-peschke) [inline tests/system/action/topic/test_json_upload.py:71]: Missing the data for headers and statistics

Comment (reiterl) [reply tests/system/action/topic/test_json_upload.py:71]: Why that? headers and statistics are not needed in the action worker.

Comment (reiterl) [reply tests/system/action/topic/test_json_upload.py:71]: I don't think, that we should include headers and statistics in the action_worker, they are just wasted load. The topic import action doesn't need them. It just needs the `rows`.

File: openslides_backend/action/actions/topic/import_.py (hunk 5 above)
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py]: @jsangmeister seems to have another opinion on this. This is a contradiction. @r-peschke  pls discuss this with @jsangmeister .

**Comment by reiterl** [comment]
Open are:

- Discussion of the overwriting of the result in the action_worker after import.
- Discussion of the headers/statistics in action worker results.

**Comment by r-peschke** [comment]
Missing:
- we should have a field `type` in the action_worker-model to differentiate between imports and real action_workers
- we should fill the `created` and `timestamp` fields on json_upload and update the `timestamp`field on import (or abort, if exists)
- on import or abort set the `state`to ènd` 
- setting state and actions still depends the open discussion points above

File: openslides_backend/action/actions/topic/import_.py (hunk 5 above)
Comment (jsangmeister) [reply openslides_backend/action/actions/topic/import_.py]: Yes, but we already discussed that a regular (e.g. daily) cleanup of all "old" import data (e.g. older than two weeks) would be the way to go here. The problem is, if we delete it during the action and the datastore request fails because of locked fields or something else, we cannot retry it. A better solution would be to include it in the `on_success` routine, but even that is not perfect because if something else goes wrong during the `import` request, the user has to regenerate the preview first before he can import again. So, for a better UX, we should keep the data here.

File: openslides_backend/action/actions/topic/json_upload.py [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+        fqid = fqid_from_collection_and_id("action_worker", self.new_store_id)
+        self.datastore.write_action_worker(
+            WriteRequest(
+                events=[
+                    Event(
+                        type=EventType.Create,
+                        fqid=fqid,
+                        fields={
+                            "id": self.new_store_id,
+                            "result": {"import": "topic", "rows": self.rows},
```
Comment (r-peschke) [inline openslides_backend/action/actions/topic/json_upload.py:105]: set the `timestamp` field of action worker to current timestamp and the status-field to `running

**Review by r-peschke — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/topic/json_upload.py (hunk 13 above)
Comment (reiterl) [reply openslides_backend/action/actions/topic/json_upload.py:107]: Add `created`, `timestamp` and `state` running to the json_upload and `timestamp` and `state` to import_
I rebased the PR.

**Review by r-peschke — CHANGES_REQUESTED** [review]

File: openslides_backend/action/actions/topic/mixins.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+        self.all_titles_in_meeting = set(
+            values.get("title")
+            for values in self.datastore.filter(
+                "topic", FilterOperator("meeting_id", "=", meeting_id), ["title"]
+            ).values()
+            if values.get("title")
+        )
+
+    def check_for_duplicate(self, title: str) -> bool:
+        result = title in self.all_titles_in_meeting
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/mixins.py:16]: I think, your code is broken. In all titles in a meeting should be every title, so I can check for it.

File: openslides_backend/action/actions/topic/mixins.py [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+
+
+class DuplicateCheckMixin(Action):
+    def init_duplicate_set(self, meeting_id: int) -> None:
+        self.all_titles_in_meeting = set(
+            values.get("title")
+            for values in self.datastore.filter(
+                "topic", FilterOperator("meeting_id", "=", meeting_id), ["title"]
+            ).values()
+            if values.get("title")
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/mixins.py:12]: Okay, removed.

File: openslides_backend/action/actions/topic/mixins.py [hunk 16]
```diff
@@ -0,0 +1,18 @@
+from ....shared.filters import FilterOperator
+from ...action import Action
+
+
+class DuplicateCheckMixin(Action):
+    def init_duplicate_set(self, meeting_id: int) -> None:
+        self.all_titles_in_meeting = set(
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/mixins.py:7]: I think a set is the right datastruct here.

File: openslides_backend/action/actions/topic/json_upload.py [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+            entry["meeting_id"] = instance["meeting_id"]
+
+        # validate and check for duplicates
+        self.init_duplicate_set(instance["meeting_id"])
+        self.rows = [self.validate_entry(entry) for entry in data]
+
+        # generate statistics
+        itemCount, itemNew, itemError = 0, 0, 0
+        for entry in self.rows:
+            itemCount += 1
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/json_upload.py:84]: Yes, that would be possible. I like more my way.

File: openslides_backend/action/actions/topic/import_.py [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+
+    def get_on_success(self, action_data: ActionData) -> Callable[[], None]:
+        def on_success() -> None:
+            for instance in action_data:
+                store_id = instance["id"]
+                self.datastore.write_action_worker(
+                    WriteRequest(
+                        events=[
+                            Event(
+                                type=EventType.Update,
```
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/import_.py]: Done.

File: openslides_backend/action/actions/topic/json_upload.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/json_upload.py]: Done. I use `len` to calculate itemCount.

File: openslides_backend/action/actions/topic/mixins.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/mixins.py:7]: I do *add* and *in* operations at the data structure.  *In* should be faster in a set and *add* is the same.

File: openslides_backend/action/actions/topic/mixins.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (reiterl) [reply openslides_backend/action/actions/topic/mixins.py:15]: Remember `result` is True, if a title is in the all_title_meeting. Then I don't need to add it., but I need to add it if it is **not** in all_titles_in_meeting.

**Review by r-peschke — CHANGES_REQUESTED** [review]

**Comment by reiterl** [comment]
Okay, add the title only if result is false. It is senseless.

**Review by r-peschke — APPROVED** [review]

File: openslides_backend/action/actions/topic/import_.py [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+                ["result"],
+                lock_result=False,
+            )
+            action_payload = [
+                entry["data"]
+                for entry in worker.get("result", {}).get("rows", [])
+                if (
+                    entry["status"] == ImportStatus.NEW
+                    or entry["status"] == ImportStatus.ERROR
+                    and entry["error"] == ["Duplicate"]
```
Comment (jsangmeister) [inline openslides_backend/action/actions/topic/import_.py:50]: Don't you import the duplicates this way too? Is this intended?

File: tests/system/action/topic/test_import.py
Comment (jsangmeister) [inline tests/system/action/topic/test_import.py:1]: I'm missing a test here with duplicates in the json upload

**Review by jsangmeister — CHANGES_REQUESTED** [review]

File: tests/system/action/topic/test_import.py
Comment (reiterl) [reply tests/system/action/topic/test_import.py:1]: I have added such a test.

**Review by jsangmeister — APPROVED** [review]

File: openslides_backend/action/actions/topic/import_.py (hunk 19 above)
Comment (jsangmeister) [reply openslides_backend/action/actions/topic/import_.py:50]: okay
